src/celeritas/user/ActionDiagnostic.cc

```diff
@@ -0,0 +1,250 @@
+    : id_(id)
+    , action_reg_(action_reg)
+    , particle_(particle)
+    , num_streams_(num_streams)
+    , store_(std::make_unique<StoreT>())
+{
+    CELER_EXPECT(id_);
+    CELER_EXPECT(action_reg_);
+    CELER_EXPECT(particle_);
+    CELER_EXPECT(num_streams > 0);
+}
+
+//---------------------------------------------------------------------------//
+//! Default destructor
+ActionDiagnostic::~ActionDiagnostic() = default;
+
+//---------------------------------------------------------------------------//
+/*!
+ * Execute action with host data.
+ */
+void ActionDiagnostic::execute(ParamsHostCRef const& params,
+                               StateHostRef& state) const
+{
+    CELER_EXPECT(params);
+    CELER_EXPECT(state);
+
+    if (!(*store_))
+    {
+        this->build_stream_store();
+    }
+    detail::tally_action(
+        params,
+        state,
+        store_->state<MemSpace::host>(state.stream_id, this->num_bins()));
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Execute action with device data.
+ */
+void ActionDiagnostic::execute(ParamsDeviceCRef const& params,
+                               StateDeviceRef& state) const
+{
+    CELER_EXPECT(params);
+    CELER_EXPECT(state);
+
+    if (!(*store_))
+    {
+        this->build_stream_store();
+    }
+    detail::tally_action(
+        params,
+        state,
+        store_->state<MemSpace::device>(state.stream_id, this->num_bins()));
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Write output to the given JSON object.
+ */
+void ActionDiagnostic::output(JsonPimpl* j) const
+{
+#if CELERITAS_USE_JSON
+    using json = nlohmann::json;
+
+    auto obj = json::object();
+    obj["actions"] = this->actions();
+    j->obj = std::move(obj);
+#else
+    (void)sizeof(j);
+#endif
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Return the diagnostic results accumulated over all streams.
+ *
+ * This builds a map of particle/action combinations to the number of
+ * occurances over all events.
+ */
+auto ActionDiagnostic::actions() const -> MapStringCount
+{
+    CELER_EXPECT(*store_);
+
+    auto const& params = store_->params<MemSpace::host>();
+
+    // Get the raw data accumulated over all host/device streams
+    auto counts = this->calc_actions();
+
+    // Map particle ID/action ID to name and store counts
+    MapStringCount result;
+    for (ActionId aid : range(ActionId{params.num_actions}))
+    {
+        for (ParticleId pid : range(ParticleId{params.num_particles}))
+        {
+            size_type bin = aid.get() * particle_->size() + pid.get();
+            CELER_ASSERT(bin < counts.size());
+
+            if (counts[bin] > 0)
+            {
+                std::string label = action_reg_->id_to_label(aid) + " "
+                                    + particle_->id_to_label(pid);
+                result[label] = counts[bin];
```

Comment:
If our primary use case for this is the demo loop/JSON output, it's probably better if we take the time to make the output a little more machine readable: perhaps the output could be an array of an arrays with a bit of metadata `"_index"  = ["actions", "particle"]`? For testing this function would still be useful though.

Reply:
Makes sense!